drivers/rtc/rtc_atmel_sam.c

```diff
@@ -0,0 +1,651 @@
+{
+	uint32_t mask = 0;
+
+	if (timalr & RTC_TIMALR_SECEN) {
+		mask |= RTC_ALARM_TIME_MASK_SECOND;
+	}
+
+	if (timalr & RTC_TIMALR_MINEN) {
+		mask |= RTC_ALARM_TIME_MASK_MINUTE;
+	}
+
+	if (timalr & RTC_TIMALR_HOUREN) {
+		mask |= RTC_ALARM_TIME_MASK_HOUR;
+	}
+
+	return mask;
+}
+
+static uint32_t rtc_atmel_sam_alarm_mask_from_calalr(uint32_t calalr)
+{
+	uint32_t mask = 0;
+
+	if (calalr & RTC_CALALR_MTHEN) {
+		mask |= RTC_ALARM_TIME_MASK_MONTH;
+	}
+
+	if (calalr & RTC_CALALR_DATEEN) {
+		mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+	}
+
+	return mask;
+}
+
+static void rtc_atmel_sam_tm_from_timalr_calalr(struct tm *timeptr, uint32_t mask,
+						 uint32_t timalr, uint32_t calalr)
+{
+	memset(timeptr, 0x00, sizeof(*timeptr));
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		timeptr->tm_sec = RTC_REG_GET_FIELD_BCD(timalr, TIMALR_SEC);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		timeptr->tm_min= RTC_REG_GET_FIELD_BCD(timalr, TIMALR_MIN);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		timeptr->tm_hour = RTC_REG_GET_FIELD_BCD(timalr, TIMALR_HOUR);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		timeptr->tm_mday = RTC_REG_GET_FIELD_BCD(calalr, CALALR_DATE);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_mon = RTC_REG_GET_FIELD_BCD(calalr, CALALR_MONTH) - 1;
+	}
+}
+
+static void rtc_atmel_sam_isr(const struct device *dev)
+{
+	struct rtc_atmel_sam_data *data = (struct rtc_atmel_sam_data *)dev->data;
+	uint32_t sr;
+
+	sr = REG_RTC_SR;
+
+	/* Update acknowledge */
+	if (sr & RTC_SR_ACKUPD) {
+		k_sem_give(&data->cr_upd_ack_sem);
+
+		REG_RTC_SCCR = RTC_SCCR_ACKCLR;
+	}
+
+	/* Alarm */
+	if ((sr & RTC_SR_ALARM) && (data->callback != NULL)) {
+		data->callback(dev, 0, data->user_data);
+
+		REG_RTC_SCCR = RTC_SCCR_ALRCLR;
+	}
+}
+
+/*************************************************************************************************/
+/*                                    RTC API implementation                                     */
+/*************************************************************************************************/
+static int rtc_atmel_sam_time_set(const struct device *dev, const struct tm *timeptr)
+{
+	struct rtc_atmel_sam_data *data = (struct rtc_atmel_sam_data *)dev->data;
+
+	uint32_t mask_supported = rtc_atmel_sam_mask_supported_get();
+
+	if (rtc_atmel_sam_tm_validate(timeptr, mask_supported) == false) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	/* Request update */
+	REG_RTC_CR = RTC_CR_UPDTIM & RTC_CR_UPDCAL;
+
+	/* Await update acknowledge */
+	k_sem_take(&data->cr_upd_ack_sem, K_MSEC(1500));
+
+	REG_RTC_TIMR = rtc_atmel_sam_timr_from_tm(timeptr);
+	REG_RTC_CALR = rtc_atmel_sam_calr_from_tm(timeptr);
+
+	REG_RTC_CR = 0;
+
+	k_mutex_unlock(&data->lock);
+
+	return 0;
+}
+
+static int rtc_atmel_sam_time_get(const struct device *dev, struct tm *timeptr)
+{
+	uint32_t timr0;
+	uint32_t calr0;
+	uint32_t timr1;
+
+	/* Validate time valid */
```

Comment:
The RTC_VER_NVTIM bit in REG_RTC_VER is set if time is invalid, so that is in fact what is happening, the comment should is there to point that out

Reply:
Them a better comment is  "_check if time and calendar are invalid_"

BTW, why are you not checking if calendar has valid data since function reads CALR?

Reply:
Oversight from my side, updated comment to state time and calendar validated, and added check of calr field in VER reg